File: 450/Utkarsh/binaryTree/preorderTraversal.cpp
//https://www.techiedelight.com/preorder-tree-traversal-iterative-recursive/

#include<bits/stdc++.h>
using namespace std;
//preorder traversal  recursive function 
void preorder(Node *root)
{
      if (root == nullptr)
        return;
    cout << root->data << " ";
    preorder(root->left);
    preorder(root->right);
}

// preorder traversal iterative funciton
void preorderIterative(Node *root)
{
    // return if tree is empty
    if (root == nullptr)
       return;
 
    // create an empty stack and push root node
    stack<Node*> stack;
    stack.push(root);
 
    // loop till stack is empty
    while (!stack.empty())
    {
        // pop a node from the stack and print it
        Node *curr = stack.top();
        stack.pop();
 
        cout << curr->data << " ";
 
        // push right child of popped node to the stack
        if (curr->right)
            stack.push(curr->right);
 
        // push left child of popped node to the stack
        if (curr->left)
            stack.push(curr->left);
 
        // important note - right child is pushed first so that left child
        // is processed first (FIFO order)
    }
}

File: 450/Ratndeep/Binary Tree/preorder traversal.py
"""
Preorder Traversal in Binary Tree 

Root -> Left Child -> RightChild 

"""

# Recursive 
def preorder(root):
    if root==None:
        return 
    print(root.data,end=" ") 
    preorder(root.left) 
    preorder(root.right) 


# Iterative 

def preorder_i(root):
    stack=[] 
    stack.append(root) 
    while stack:
        top = stack.pop()
        print(top,end=" ") 
        if top.right:
            stack.append(top.right) 
        if top.left:
            stack.append(top.left)

class Node:
    def __init__(self,data):
        self.data=data 
        self.left=None 
        self.right=None 

class BTree:
    def __init__(self):
        self.head=None 

    def addNode(self,data): 
        node = Node(data) 
        root=self.head
        if root==None:
            self.head=node 
        else:
            while(root):
                if root.data<node.data:
                    if root.right==None:
                        root.right = node 
                        break 
                    else: 
                        root=root.right 
                else: 
                    if root.left==None: 
                        root.left=node 
                        break 
                    else: 
                        root=root.left 
        
    

bt = BTree() 
data = list(map(int,input().split()))
for i in range(len(data)):
    bt.addNode(data[i])  
preorder_i(bt.head)